x-pack/metricbeat/module/azure/monitor/client.go

```diff
@@ -0,0 +1,238 @@
+			if err == nil {
+				for _, res := range resourceList {
+					for _, metric := range resource.Metrics {
+						met, err := client.mapMetric(metric, res)
+						if err != nil {
+							return err
+						}
+						metrics = append(metrics, met)
+					}
+				}
+			}
+
+		}
+		if resource.Query != "" {
+			resourceList, err := client.azureMonitorService.GetResourcesByResourceQuery(resource.Query)
+			if err == nil {
+				for _, res := range resourceList {
+					for _, metric := range resource.Metrics {
+						met, err := client.mapMetric(metric, res)
+						if err != nil {
+							return err
+						}
+						metrics = append(metrics, met)
+					}
+				}
+			}
+		}
+	}
+	client.resources.metrics = metrics
+	return nil
+}
+
+// GetMetricValues returns the specified metric data points for the specified resource ID/namespace.
+func (client *Client) GetMetricValues() error {
+	for i, metric := range client.resources.metrics {
+		client.resources.metrics[i].values = nil
+		endTime := time.Now().UTC()
+		startTime := endTime.Add(client.config.Period * (-1))
+		timespan := fmt.Sprintf("%s/%s", startTime.Format(time.RFC3339), endTime.Format(time.RFC3339))
+		//interval := "PT1M"
+		//interval=""
+		var filter string
+		if len(metric.dimensions) > 0 {
+			var filterList []string
+			for _, dim := range metric.dimensions {
+				filterList = append(filterList, dim.name+" eq '"+dim.value+"'")
+			}
+			filter = strings.Join(filterList, " AND ")
+		}
+		resp, err := client.azureMonitorService.GetMetricValues(metric.resource.ID, metric.namespace, timespan, metric.names,
+			metric.aggregations, filter)
+		if err != nil {
+			return err
+		}
+		for _, v := range resp {
+			for _, t := range *v.Timeseries {
+				for _, mv := range *t.Data {
+					var val MetricValue
+					val.name = *v.Name.Value
+					val.timestamp = mv.TimeStamp.Time
+					if mv.Minimum != nil {
+						val.min = mv.Minimum
+					}
+					if mv.Maximum != nil {
+						val.max = mv.Maximum
+					}
+					if mv.Average != nil {
+						val.average = mv.Average
+					}
+					if mv.Total != nil {
+						val.total = mv.Total
+					}
+					if mv.Count != nil {
+						val.count = mv.Count
+					}
+					client.resources.metrics[i].values = append(client.resources.metrics[i].values, val)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// mapMetric should validate and map the metric related configuration to relevant azure monitor api parameters
+func (client *Client) mapMetric(metric azure.MetricConfig, resource resources.GenericResource) (Metric, error) {
+	var dim []Dimension
+	var supportedMetricNames []string
+	var unsupportedMetricNames []string
+	var supportedAggregations []string
+	var unsupportedAggregations []string
+	metricDefinitions, err := client.azureMonitorService.GetMetricDefinitions(*resource.ID, metric.Namespace)
+	if err != nil {
+		client.log.Errorf("No metric definitions were found for resource %s and namespace %s. Error %v", *resource.ID, metric.Namespace, err)
+		return Metric{}, err
```

Comment:
Would be better if you could return `nil` here, meaning it would need type `*Metric`. Not sure if possible / reasonable.